bin/echo.c

```diff
@@ -253,17 +254,25 @@ int negotiate(struct s2n_connection *conn, int fd)
     return 0;
 }
 
-int renegotiate(struct s2n_connection *conn, int fd)
+int renegotiate(struct s2n_connection *conn, int fd, bool wait_for_more_data)
 {
     s2n_blocked_status blocked = S2N_NOT_BLOCKED;
     uint8_t buffer[STDIO_BUFSIZE] = { 0 };
     ssize_t data_read = 0;
 
     GUARD_RETURN(s2n_renegotiate_wipe(conn), "Unable to prepare connection for renegotiate");
+    GUARD_RETURN(s2n_connection_set_client_auth_type(conn, S2N_CERT_AUTH_OPTIONAL), "Error setting ClientAuth optional");
 
     fprintf(stdout, "RENEGOTIATE\n");
     fflush(stdout);
 
+    /* Do not proceed with renegotiation until we receive more data from the server */
+    if (wait_for_more_data) {
+        fd_set fds = { 0 };
+        FD_SET(fd, &fds);
+        select(FD_SETSIZE, &fds, NULL, NULL, NULL);
+    }
```

Comment:
So this works because `s2n_renegotiate` calls `s2n_negotiate`, which tries to send the client hello, but can't because you set `fd` to be read only? How does `fd` become writeable again?

Reply:
I'm not setting fd to read only. [`select`](https://www.man7.org/linux/man-pages/man2/select.2.html) will block until fd becomes ready for read, meaning that data exists to read because the server has sent more data.

In practice it works basically the same if you swap out `select` for `sleep`.

Reply:
Ohh I see, thanks!